src/mon/Elector.cc

```diff
@@ -24,108 +24,103 @@
-  mon->store->apply_transaction(t);
+  t->put(Monitor::MONITOR_NAME, "election_writeable_test", rand());
+  int r = mon->store->apply_transaction(t);
+  ceph_assert(r >= 0);
+}
 
-  mon->join_election();
+bool Elector::is_current_member(int rank)
+{
+  return mon->quorum.count(rank);
+}
 
-  // clear up some state
-  electing_me = false;
-  acked_me.clear();
+void Elector::trigger_new_election()
+{
+  mon->start_election();
 }
 
+int Elector::get_my_rank()
+{
+  return mon->rank;
+}
 
-void Elector::start()
+void Elector::reset_election()
 {
-  if (!participating) {
-    dout(0) << "not starting new election -- not participating" << dendl;
-    return;
-  }
-  dout(5) << "start -- can i be leader?" << dendl;
+  mon->bootstrap();
+}
 
-  acked_me.clear();
-  init();
-  
-  // start by trying to elect me
-  if (epoch % 2 == 0) {
-    bump_epoch(epoch+1);  // odd == election cycle
-  } else {
-    // do a trivial db write just to ensure it is writeable.
-    auto t(std::make_shared<MonitorDBStore::Transaction>());
-    t->put(Monitor::MONITOR_NAME, "election_writeable_test", rand());
-    int r = mon->store->apply_transaction(t);
-    ceph_assert(r >= 0);
-  }
-  electing_me = true;
-  acked_me[mon->rank].cluster_features = CEPH_FEATURES_ALL;
-  acked_me[mon->rank].mon_release = ceph_release();
-  acked_me[mon->rank].mon_features = ceph::features::mon::get_supported();
-  mon->collect_metadata(&acked_me[mon->rank].metadata);
-  leader_acked = -1;
+bool Elector::ever_participated()
+{
+  return mon->has_ever_joined;
+}
+
+unsigned Elector::paxos_size()
+{
+  return (unsigned)mon->monmap->size();
+}
+
+void Elector::shutdown()
+{
+  cancel_timer();
+}
+
+void Elector::notify_bump_epoch()
+{
+  mon->join_election();
+}
 
+void Elector::propose_to_peers(epoch_t e)
+{
   // bcast to everyone else
   for (unsigned i=0; i<mon->monmap->size(); ++i) {
     if ((int)i == mon->rank) continue;
     MMonElection *m =
-      new MMonElection(MMonElection::OP_PROPOSE, epoch, mon->monmap);
+      new MMonElection(MMonElection::OP_PROPOSE, e, mon->monmap);
     m->mon_features = ceph::features::mon::get_supported();
     m->mon_release = ceph_release();
     mon->send_mon_message(m, i);
-  }
-  
-  reset_timer();
+  }  
 }
 
-void Elector::defer(int who)
+void Elector::_start()
 {
-  dout(5) << "defer to " << who << dendl;
-
-  if (electing_me) {
-    // drop out
-    acked_me.clear();
-    electing_me = false;
-  }
+  peer_info.clear();
+  peer_info[mon->rank].cluster_features = CEPH_FEATURES_ALL;
+  peer_info[mon->rank].mon_release = ceph_release();
+  peer_info[mon->rank].mon_features = ceph::features::mon::get_supported();
+  mon->collect_metadata(&peer_info[mon->rank].metadata);
+  reset_timer();
+}
 
-  // ack them
-  leader_acked = who;
-  MMonElection *m = new MMonElection(MMonElection::OP_ACK, epoch, mon->monmap);
+void Elector::_defer_to(int who)
+{
+  MMonElection *m = new MMonElection(MMonElection::OP_ACK, get_epoch(), mon->monmap);
```

Comment:
nit, could use `auto`. less repeating.

Reply:
Maybe next time; this originated in old code.